scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -440,32 +505,225 @@ static void refresh_tune_params(void)
+	 * have a disproporionate impact on the latency (and possibly
+	 * throughput) of the workload they are enabling due to Amdahl's law.
+	 * For example, say that you have a workload where 50% of the workload
+	 * is serialized by a producer and consumer task (25% each), and the
+	 * latter 50% is serviced in parallel by n CPU hogging tasks. If either
+	 * the producer or consumer is improved by a factor of x, it improves
+	 * the latency of the entire workload by:
+	 *
+	 *	S_lat(x) = 1 / ((1 - .25) + (.25 / x))
+	 *
+	 * Say that we improve wakeup latency by 2x for either task, the
+	 * latency improvement would be:
+	 *
+	 *	S_lat(2) = 1 / ((1 - .25) + (.25 / 2))
+	 *		 = 1 / (.75 + .125)
+	 *		 = 1 / .875
+	 *		~= 14.2%
+	 *
+	 * If we instead improve wakeup latency by 2x for all the n parallel
+	 * tasks in the latter 50% of the workload window, the improvement
+	 * would be:
+	 *
+	 *	S_lat(2) = 1 / ((1 - .5) + (.5 / 2))
+	 *		 = 1 / (.5 + .25)
+	 *		 = 1 / .75
+	 *		~= 33%
+	 *
+	 * This is also significant, but the returns are amortized across all
+	 * of those tasks. Thus, by giving a latency boost to the producer /
+	 * consumer tasks, we optimize for the case of scheduling tasks that
+	 * are on the critical path for serial workchains, and have a
+	 * disproporionate impact on the latency of a workload.
+	 *
+	 * We multiply the frequencies of wait_freq and waker_freq somewhat
+	 * arbitrarily, based on observed performance for audio and gaming
+	 * interactive workloads.
+	 */
+	waker_freq = max_clamp(taskc->waker_freq, DL_FREQ_FT_MAX);
+	blocked_freq = max_clamp(taskc->blocked_freq, DL_FREQ_FT_MAX);
+	freq_factor = blocked_freq * waker_freq * waker_freq;
+	freq_factor = scale_fair(freq_factor, p->scx.weight);
+
+	/*
+	 * The above frequencies roughly follow an exponential distribution, so
+	 * borrow the bpf_log2l() implementation from lavd to linearize it to a
+	 * boost priority.
+	 */
+	lat_prio = bpf_log2l(freq_factor + 1);
+	lat_prio = max_clamp(lat_prio, DL_MAX_LAT_PRIO);
+
+	/*
+	 * Next calculate a task's average runtime, and apply it to deadline
+	 * accordingly. A task with a large runtime is penalized from an
+	 * interactivity standpoint, for obvious reasons.
+	 *
+	 * As with waker and blocked frequencies above, this follows an
+	 * exponential distribution. We inversely scale to account for
+	 * empirical observations which seem to bring it roughly to the same
+	 * order of magnitude as the blocker and waker frequencies above.
+	 */
+	avg_run_raw = taskc->avg_runtime / DL_RUNTIME_SCALE;
+	avg_run_raw = max_clamp(avg_run_raw, DL_MAX_LATENCY_NS);
+	avg_run_raw = scale_fair(avg_run_raw, p->scx.weight);
+	avg_run = bpf_log2l(avg_run_raw + 1);
+
+	if (avg_run < lat_prio) {
+		lat_prio -= avg_run;
+	} else {
+		lat_prio = 0;
+	}
+
+	lat_scale = sched_prio_to_latency_weight(lat_prio);
+	lat_scale = max_clamp(lat_scale, LB_MAX_WEIGHT);
+
+	return scale_fair(taskc->avg_runtime, lat_scale);
```

Comment:
It will be nice to describe the idea behind this logic. As far as I understand, essentially the task's deadline is determined by its average runtime scaled down by weighted latency priority, which is again determined by two frequencies and runtime.

Reply:
I've added a comment that hopefully captures what's going on here a bit more clearly. But yes, you're exactly right. Essentially what we're doing is:

1. Computing some "lat_prio" value as a function of `blocked_freq`, `waker_freq`, and `avg_runtime`; all scaled by a task's weight. `blocked_freq` and `waker_freq` scale `lat_prio` positively, and `avg_runtime` scales `lat_prio` negatively.
2. With `lat_prio`, we compute the weight value that will ultimately be used to calculate our deadline value (relative to a task's current vtime). Let's call this weight `wi_lat`
3. We finally compute the task's deadline (i.e. CPU request length r_i, in EEVDF terms) by computing the inverse scaling of its average runtime according to `wi_lat`:

```
r_i = taskc->avg_runtime * 100 / wi_lat
```